seed/seed20.go

```diff
@@ -0,0 +1,348 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seed
+
+/* ATTN this should *not* use:
+
+* dirs package: it is passed an explicit directory to work on
+
+* release.OnClassic: it assumes classic based on the model classic
+  option; consistency between system and model can/must be enforced
+  elsewhere
+
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/snapasserts"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/seed/internal"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/timings"
+)
+
+type seed20 struct {
+	systemDir string
+
+	db asserts.RODatabase
+
+	model *asserts.Model
+
+	snapDeclsByID   map[string]*asserts.SnapDeclaration
+	snapDeclsByName map[string]*asserts.SnapDeclaration
+
+	snapRevsByID map[string]*asserts.SnapRevision
+
+	auxInfos map[string]*internal.AuxInfo20
+
+	snaps             []*Snap
+	essentialSnapsNum int
+}
+
+func (s *seed20) LoadAssertions(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+	if db == nil {
+		// a db was not provided, create an internal temporary one
```

Comment:
what's the use case of passing a nil db in?
asking because at first I thought sometimes you'd pass a nil `db` in with `commitTo` doing all the work, but if `db` is nil `commitTo` is overridden so :man_shrugging:

Reply:
it takes the two argument because one is a read-only facet, and the other is a very specific capability to write to the underlying db.  nils are supported for code that wants to load a seed in a one off way and doesn't need to keep assertions etc in long-lived db, Pawel new stuff is the first actual user, but there are other UC20 tooling bits that will need the same:

https://github.com/snapcore/snapd/pull/7658/files#diff-eb9825aa18d9bbbcc41ca31728af8157R75